### PR TITLE
feat: remove steps from config [DET-3298, DET-3299, DET-3525]

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change allows all searchers and `min_validation_period` and `min_checkpoint_period` to be configured in terms of records, batches or epochs. It also modifies all searchers to emit operations, such as `Train{Epochs: 2}` allowing searchers to be oblivious to how requested work is broken into workloads. Instead, the trial workload sequencer handles this and relays completed operations back to the searcher. It still relays workloads back to the experiment, though this is only to check the status of validations (is it the best?) and to record units complete so that progress is more granular.

## Test Plan
- [x] run chaos tests for all searchers, while pausing and restarting the master each steps.
- [x] write tests for new sequencer
- [x] updates all tests for all searchers
- [x] run examples
- [x] make sure old experiments still load, but it's ok that they can't be forked.

## Commentary
Because the searcher is "workload unaware", the experiment also needs to be. Because of this, some other workload related tasks, like caching unrequested checkpoints until they are needed, are also now the responsibility of the trial workload sequencer. This has slimmed down the event log considerably and should allow the event log and search method to be decoupled easier in follow up tech debt work.

Nice things about writing it this way:
* searchers, `min_validation_period` and `min_checkpoint_period` can all be configure in terms of different units.
* the expected behavior of `min_validation_period` and `min_checkpoint_period` is 👌 
* the trial/experiment interact is simpler.
* searchers are way simpler than any other way this could be written.
* it makes test set metrics, initial validations little work in the new sequencer and mostly just harness side work.

Allow unfortunately, there is basically no way to make this PR any smaller, or to land it in small pieces.

### CODE REVIEW

File: master/internal/db/postgres.go [hunk 1]
```diff
@@ -1119,13 +1119,37 @@ WHERE id = :id`, setClause(toUpdate)), trial)
...
+}
+
 // RollBackTrial deletes from the database all steps, checkpoints, and validations for the trial
 // that correspond to steps past lastStep.
 func (db *PgDB) RollBackTrial(id int, lastStep int) error {
 	// This delete cascades to checkpoints and validations.
 	_, err := db.sql.Exec(`
 DELETE FROM steps
 WHERE trial_id = $1 AND id > $2
+   OR trial_id = $1 AND id = $2 AND state != 'COMPLETED'
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/db/postgres.go:1152]: In the event we checkpoint, then validate but don't finish it, when we rollback the sequencer it will say to roll back to the end up that step (`id > $2`) and this query deletes all steps after that, but we still have a started but not completed validation in the same step (`id = $2`). There isn't really any harm in this but in master restart you see a bunch of `failed to save workload to the database/duplicate val` in `processAssigned` because the validation never got cleared and when we get assigned we start it over and save it. This is reproducible on latest master, too.

File: master/internal/db/postgres.go [hunk 2]
```diff
@@ -1119,13 +1119,37 @@ WHERE id = :id`, setClause(toUpdate)), trial)
 	return nil
 }
 
+// RollbackSearcherEvents rolls back the events for an experiment to the last step with a
+// checkpoint.
+func (db *PgDB) RollbackSearcherEvents(experimentID int) error {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/db/postgres.go:1124]: Sure to the comment, and what about `RollbackSearcherEventsToLastStepWithCheckpoint`. Really long but precise.

File: master/internal/trial_workload_sequencer.go [hunk 3]
```diff
@@ -1,63 +1,103 @@
...
+	curOpIdx  int
+	curStepID int
+
+	latestCheckpoint *model.Checkpoint
+	// There are some strange corner cases where the sequencer will receive checkpoint completed
+	// messages before it requests them in the event a trial is descheduled or paused. For example,
+	// run a trial with a step that will validate and checkpoint, pause it on that step then unpause
+	// the trial and restart the master. The events will be replayed to the master and it received a
+	// checkpoint completed while waiting for a validation and throw the checkpoint away as
+	// unexpected without this.
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial_workload_sequencer.go:59]: I wrote all these comments really quickly, I'll reword it, ha.

File: master/internal/trial_workload_sequencer_test.go [hunk 4]
```diff
@@ -136,11 +148,10 @@ checkpoint_policy: none
...
 
 	// Request a few operations so the sequencer builds its internal desired state.
-	assert.NilError(t, s.OperationRequested(trainOperation1))
+	assert.NilError(t, s.OperationRequested(train))
 	assert.Assert(t, !s.UpToDate())
-	assert.NilError(t, s.OperationRequested(trainOperation2))
-	assert.NilError(t, s.OperationRequested(checkpointOperation2))
-	assert.NilError(t, s.OperationRequested(validateOperation2))
+	assert.NilError(t, s.OperationRequested(validate))
+	assert.NilError(t, s.OperationRequested(checkpoint))
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial_workload_sequencer_test.go:154]: Any searcher that requests checkpoints (PBT) still requests them in this order. The way I went about the "searchers always do a preval checkpoint" is to add that logic into the workload sequencer, mainly so if we add another searcher, it doesn't get forgotten.

File: master/internal/trial_workload_sequencer.go [hunk 5]
```diff
@@ -74,234 +114,338 @@ func (s *trialWorkloadSequencer) WorkloadManagerType() model.WorkloadManagerType
...
+
+	if s.minValidationNeeded() {
+		return s.validate(), nil
+	}
+
+	if s.minCheckpointNeeded() {
+		return s.checkpoint(), nil
+	}
+
+	switch tOp := s.ops[s.curOpIdx].(type) {
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial_workload_sequencer.go:308]: typed operation, some of the `tOp := op.(type)`, because it most placed I do this, it's something like `tOp := op.(type)`. I'll change this though.

File: master/internal/trial_workload_sequencer.go [hunk 6]
```diff
@@ -74,234 +114,338 @@ func (s *trialWorkloadSequencer) WorkloadManagerType() model.WorkloadManagerType
...
+	case searcher.Validate:
+		// We choose to always checkpoint before completing any searcher operations. This allows us
+		// to confidently rollback searcher events and all trials while keeping them in a consistent
+		// state.
+		if s.batchesSinceLastCkpt != 0 {
+			return s.checkpoint(), nil
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/trial_workload_sequencer.go:314]: Can't throw an error here because this is what actually manages inserting the checkpoint. Searchers are unaware of it.

File: master/internal/db/postgres.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/internal/db/postgres.go:1124]: If I had to pick a detail to put in the name, I would name it `RollbackSearchEventsForMasterRestart` or something.  To me it is more important to explain that it belongs to the super-special master restart codepath than to explain precisely what it is doing.

After reconsidering though, I think having the details in the comment is much more important than renaming the function.

File: master/internal/db/postgres.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/internal/db/postgres.go:1152]: Ok, cool.

File: master/internal/db/postgres.go [hunk 7]
```diff
@@ -1119,13 +1119,37 @@ WHERE id = :id`, setClause(toUpdate)), trial)
...
+        (se.content->'msg'->'workload'->>'trial_id')::int trial_id,
+        max(CASE WHEN content->'msg'->'workload'->>'kind' = 'CHECKPOINT_MODEL'
+        THEN (se.content->'msg'->'workload'->>'step_id')::int ELSE 0 END) latest_checkpoint_id
+    FROM searcher_events se
+    GROUP BY trial_id
+	) lc
+WHERE experiment_id = $1
+    AND (se.content->'msg'->'workload'->>'trial_id')::int = lc.trial_id
+    AND (se.content->'msg'->'workload'->>'step_id')::int > lc.latest_checkpoint_id;
+	`, experimentID)
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/db/postgres.go:1138]: Miss this by accident because it was hidden in the middle (damn github, I dont like that feature). I'll leave a good comment. Query surprisingly isn't terrible performance wise. I was trying instead for something more like 
```
DELETE FROM searcher_events se
USING (
    SELECT
        (se.content->'msg'->'workload'->>'trial_id')::int trial_id,
        first(content->'msg'->'workload'->>'kind' = 'CHECKPOINT_MODEL')::int step_id
    FROM searcher_events se
	GROUP BY trial_id
	ORDER BY (se.content->'msg'->'workload'->>'step_id')::int DESC
	) latest_checkpoint
WHERE experiment_id = 1
    AND (se.content->'msg'->'workload'->>'trial_id')::int = lclatest_checkpoint.trial_id
    AND (se.content->'msg'->'workload'->>'step_id')::int > latest_checkpoint.step_id;
```
with an index like
```
CREATE INDEX ix_searcher_events_exp_trial_step_id ON public.searcher_events USING btree (
  experiment_id,
  ((searcher_events.content->'msg'->'workload'->>'trial_id')::int),
  ((searcher_events.content->'msg'->'workload'->>'step_id')::int)
);
```
but I can't quite get it to work (postgres complaining about bad usage of `first`, I guess I dont know how the first agg function works).

File: master/internal/db/postgres.go [hunk 8]
```diff
@@ -1119,13 +1119,37 @@ WHERE id = :id`, setClause(toUpdate)), trial)
...
+	_, err := db.sql.Exec(`
+DELETE FROM searcher_events se
+USING (
+    SELECT
+        (se.content->'msg'->'workload'->>'trial_id')::int trial_id,
+        max(CASE WHEN content->'msg'->'workload'->>'kind' = 'CHECKPOINT_MODEL'
+        THEN (se.content->'msg'->'workload'->>'step_id')::int ELSE 0 END) latest_checkpoint_id
+    FROM searcher_events se
+    GROUP BY trial_id
+	) lc
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/db/postgres.go:1134]: `latest_checkpoint`, i'll rename

File: master/pkg/searcher/hyperparameters.go [hunk 9]
```diff
@@ -10,6 +10,17 @@ import (
 
 type hparamSample map[string]interface{}
 
+func (h hparamSample) GlobalBatchSize() int {
+	// If the hyperparameters.global_batch_size is configured as a const hyperparameter,
+	// we infer its type to be a float but in some cases, its type can be specified and an
+	// int is also valid.
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/searcher/hyperparameters.go:16]: This isn't because someone chose a float but because when golang sees a numeric value and is told to deserialize it into an `interface{}` like in:
```
// ConstHyperparameter is a constant.
type ConstHyperparameter struct {
	Val interface{} `json:"val"`
}
```
it _always_ chooses a float64:

https://play.golang.org/p/ridKntaeRMp

File: master/pkg/searcher/message.go [hunk 10]
```diff
@@ -8,7 +8,7 @@ import (
 	"github.com/pkg/errors"
 )
 
-// CompletedMessage is the wrapping message returned by the trial runner when a workload
+// CompletedMessage is the wrapping message returned by the trial runner when a workload`
```
Comment (stoksc) [inline master/pkg/searcher/message.go:11]:
```suggestion
// CompletedMessage is the wrapping message returned by the trial runner when a workload
```

File: master/pkg/searcher/search_method.go [hunk 11]
```diff
@@ -22,51 +22,52 @@ type SearchMethod interface {
 	trialCreated(ctx context, requestID RequestID) ([]Operation, error)
 	// trainCompleted informs the searcher that the training workload initiated by the same searcher
 	// has completed. It returns any new operations as a result of this workload completing.
-	trainCompleted(ctx context, requestID RequestID, message Workload) ([]Operation, error)
+	trainCompleted(ctx context, requestID RequestID, train Train) ([]Operation, error)
```
[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/searcher/search_method.go:25]: We talked about this offline. The experiment actor is still relayed workload completed messages for the sole purpose of checking if a validation was the best and recording how many units the trial is completing so that the above doesn't kill searcher granularity.

File: master/pkg/searcher/simulate.go [hunk 12]
```diff
@@ -21,24 +21,24 @@ func ConstantValidation(_ *rand.Rand, _, _ int) float64 { return 1 }
...
 	summary := make(map[string]int)
 
 	for _, workloads := range s {
 		key := ""
-		for _, workloadMsg := range workloads {
-			switch kind := workloadMsg.Workload.Kind; kind {
-			case RunStep:
+		for _, workload := range workloads {
+			switch workload.(type) {
+			case Train:
```
Comment (stoksc) [inline master/pkg/searcher/simulate.go:34]: This is incorrect but gets fixed by #941.

File: master/pkg/searcher/workload.go [hunk 13]
```diff
@@ -31,9 +31,10 @@ type Workload struct {
...
 func (w Workload) String() string {
-	extra := ""
+	var extra string
 	if w.Kind == RunStep {
-		extra = fmt.Sprintf(" (%d Batches)", w.NumBatches)
+		extra += fmt.Sprintf(" (%d Batches)", w.NumBatches)
 	}
+	extra += fmt.Sprintf(" (%d Prior Batches)", w.TotalBatchesProcessed)
 	return fmt.Sprintf("<%s%s: (%d,%d,%d)>", w.Kind, extra, w.ExperimentID, w.TrialID, w.StepID)
 }
```
Comment (stoksc) [inline master/pkg/searcher/workload.go:40]: TODO later, move this file to `master/internal` since it belongs to the workload sequencer now.

File: e2e_tests/tests/fixtures/no_op/single-in-epochs.yaml [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  metrics_sigma: 0
+searcher:
+  metric: validation_error
+  smaller_is_better: true
+  name: single
+  max_length:
+    epochs: 2
+reproducibility:
+  experiment_seed: 999
+records_per_epoch: 48000
```
[the remark this answers is not in the document]
Comment (stoksc) [reply e2e_tests/tests/fixtures/no_op/single-in-epochs.yaml:20]: (this is really only 15 steps of a no-op, but that's still a lot)

File: e2e_tests/tests/fixtures/no_op/single-in-epochs.yaml [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  metrics_sigma: 0
+searcher:
+  metric: validation_error
+  smaller_is_better: true
+  name: single
+  max_length:
+    epochs: 2
+reproducibility:
+  experiment_seed: 999
+records_per_epoch: 8000
```
Comment (rb-determined-ai) [inline e2e_tests/tests/fixtures/no_op/single-in-epochs.yaml:20]: any particular reason why these need to be so long?

File: e2e_tests/tests/fixtures/no_op/single-in-records.yaml [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  global_batch_size: 32
+  metrics_progression: decreasing
+  metrics_base: 0.9
+  metrics_sigma: 0
+searcher:
+  metric: validation_error
+  smaller_is_better: true
+  name: single
+  max_length:
+    records: 8000
```
Comment (rb-determined-ai) [inline e2e_tests/tests/fixtures/no_op/single-in-records.yaml:17]: ditto

File: master/pkg/model/length.go [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+	switch ctx.defaultUnit {
+	case Records:
+		return NewLengthInRecords(batches * ctx.globalBatchSize)
+	case Batches:
+		return NewLengthInBatches(batches)
+	case Epochs:
+		numEpochs := math.Ceil(float64(batches*ctx.globalBatchSize) / float64(ctx.recordsPerEpoch))
+		return NewLengthInEpochs(int(numEpochs))
+	default:
+		panic(fmt.Sprintf("invalid unit in ctx: %s", ctx.defaultUnit))
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/pkg/model/length.go:119]: ok

File: master/pkg/searcher/event_log.go [hunk 18]
```diff
@@ -17,66 +21,41 @@ type TrialClosedEvent struct {
...
 // EventLog records all actions coming to and from a searcher.
 type EventLog struct {
-	uncommitted []Event
-
-	inFlightWorkloads map[WorkloadOperation]bool
-
-	// completedWorkloads contains a bool for each WorkloadOperation which has been completed. The
-	// value indicates whether EventLog.WorkloadCompleted has returned true for that
-	// WorkloadOperation, which must only happen once for each distinct operation. The operation is
-	// logged the first time it is seen, regardless.
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply master/pkg/searcher/event_log.go:27]: If we don't ever see duplicate events... can we just delete this member entirely?

File: e2e_tests/tests/fixtures/no_op/single-in-epochs.yaml (hunk 15 above)
Comment (stoksc) [reply e2e_tests/tests/fixtures/no_op/single-in-epochs.yaml:20]: `(2 * 8000 records) / (32 records/batch * 100 batches/step)` puts this at 5 steps, which I think is pretty reasonable.

File: e2e_tests/tests/fixtures/no_op/single-in-records.yaml (hunk 16 above)
Comment (stoksc) [reply e2e_tests/tests/fixtures/no_op/single-in-records.yaml:17]: This is 2 1/2 steps.

File: master/pkg/searcher/event_log.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply master/pkg/searcher/event_log.go:27]: yep, I just needed some time to think through it.
